ml-algorithms/src/main/java/org/opensearch/ml/engine/utils/AgentModelsSearcher.java

```diff
@@ -0,0 +1,47 @@
+package org.opensearch.ml.engine.utils;
+
+import static org.opensearch.ml.common.CommonValue.ML_AGENT_INDEX;
+import static org.opensearch.ml.common.CommonValue.TOOL_PARAMETERS_PREFIX;
+
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.index.query.BoolQueryBuilder;
+import org.opensearch.index.query.QueryBuilders;
+import org.opensearch.ml.common.spi.tools.Tool;
+import org.opensearch.ml.common.spi.tools.WithModelTool;
+import org.opensearch.search.builder.SearchSourceBuilder;
+
+public class AgentModelsSearcher {
+    private final Set<String> relatedModelIdSet;
+
+    public AgentModelsSearcher(Map<String, Tool.Factory> toolFactories) {
+        relatedModelIdSet = new HashSet<>();
+        for (Map.Entry<String, Tool.Factory> entry : toolFactories.entrySet()) {
+            Tool.Factory toolFactory = entry.getValue();
+            if (toolFactory instanceof WithModelTool.Factory) {
+                WithModelTool.Factory withModelTool = (WithModelTool.Factory) toolFactory;
+                relatedModelIdSet.addAll(withModelTool.getAllModelKeys());
+            }
+        }
+    }
+
+    /**
+     * Construct a should query to search all agent which containing candidate model Id
+    
+     @param candidateModelId the candidate model Id
+     @return a should search request towards agent index.
+     */
+    public SearchRequest constructQueryRequestToSearchModelIdInsideAgent(String candidateModelId) {
```

Comment:
Can we filter out hidden agent here as well, as we are not doing anything with hidden model. So we should not search for hidden agent too.

Reply:
Sure. Let me change it.

Reply:
I add must query when we create query in agentModelSearcher to make sure its is_hidden is false. Also remove  UT about hidden model, Please check it.

Reply:
what if is_hidden field is null or empty, I meant the version when we didn't introduce hidden_model feature.

Reply:
Good point! Now I tune the DSL and if an agent doesn't have is_hidden field and use candidate model, we will also block it.